### PR TITLE
gh-106213: Make Emscripten trampolines work with JSPI

### PR DESCRIPTION
**Pull request by hoodmane**
There is a WIP proposal to enable webassembly stack switching which have been implemented in v8:

https://github.com/WebAssembly/js-promise-integration

It is not possible to switch stacks that contain JS frames so the Emscripten JS trampolines that allow calling functions with the wrong number of arguments don't work in this case. However, the js-promise-integration proposal requires the [type reflection for Wasm/JS API](https://github.com/WebAssembly/js-types) proposal, which allows us to actually count the number of arguments a function expects.

For better compatibility with stack switching, this PR checks if type reflection is available, and if so we use a switch block to decide the appropriate signature. If type reflection is unavailable, we should use the current EMJS trampoline.

We cache the function argument counts since when I didn't cache them performance was negatively affected.

A backport of this patch to v3.11.x is tested against Pyodide and passes our tests:
https://github.com/pyodide/pyodide/pull/3964

* Issue: gh-106213

### CODE REVIEW

**Review by erlend-aasland — APPROVED** [review]
Build changes LGTM; I cannot comment on the Emscripten peculiarities, though :)

**Comment by hoodmane** [comment]
I guess I will add a news entry.

**Comment by hoodmane** [comment]
Thanks for the review @erlend-aasland! Emscripten is indeed peculiar...

**Comment by erlend-aasland** [comment]
> Thanks for the review @erlend-aasland! Emscripten is indeed peculiar...

If you have the time to guide me, I'll be happy to review the rest of the PR, just for the fun of learning! :)

File: Python/emscripten_trampoline.c
Comment (hoodmane) [inline Python/emscripten_trampoline.c]: We need to ensure that this `int` is exported so we can set it from the `_PyEM_CountFuncParams` initialization logic.

File: Python/emscripten_trampoline.c
Comment (hoodmane) [inline Python/emscripten_trampoline.c]: `EM_JS` defines a C signature with a JavaScript body. `EM_JS_DEPS` declares a runtime dependency of an `EM_JS` function on Emscripten JS runtime functions. Both work by injecting strings into custom sections of the generated object file. These sections are inspected and removed by the linker.

File: Python/emscripten_trampoline.c
Comment (hoodmane) [inline Python/emscripten_trampoline.c]: `WebAssembly.Function.type` is the JS type reflection API. It returns an object like:
```js
{ parameters: ['i32', 'i32', 'i32'], results: ['i32'] }
```
This would be for a PyCFunctionWithKeywords function. Wasm has 32 bit pointers, so `PyObject*` is just 'i32'. Currently the type are: i32, i64, v128, f32, f64, and externref (an externref is an opaque reference to a JS object).

We assume that the function actually takes 0, 1, 2, or 3 parameters which are all i32s. So we just need to count how many there are.

Frequent calls to `WebAssembly.Function.type` are expensive so we cache them.

File: Python/emscripten_trampoline.c
Comment (hoodmane) [inline Python/emscripten_trampoline.c]: This is now initialization code. It runs very early in the startup of the Emscripten loader logic. We want to make the signature cache and set `_PyEM_type_reflection_available` to the correct value. We can feature detect type reflection just with `"Function" in WebAssembly`.

It's too early to access WebAssembly memory so we use the `preRun` hook to set it after everything is ready but before `_start` is called. 

`Module.preRun` is either `undefined`, a function, or a list of functions. We ensure that it's a list of functions and then push a function that does our instantiation. `__PyEM_type_reflection_available` contains the pointer to `_PyEM_type_reflection_available` which we need to convert into an index into the HEAP32 by dividing by 4.

Comment (hoodmane) [reply Python/emscripten_trampoline.c]: It would probably be better to define a `PyEM_trampoline_init` function which would be called during interpreter initialization and then `_PyEM_type_reflection_available` could be static. If you can tell me where in the interpreter initialization / preinitialization such a function should be called, I can switch to that. The `PyEM_trampoline_init` function would be idempotent and have no teardown, it just needs to be called at least once.

**Review by hoodmane — COMMENTED** [review]
Here's the background of these trampolines and why this PR.

### The purpose of these trampolines

Each WebAssembly function has a signature known to the runtime: how many arguments of what types and how many return values of what types. For security reasons, WebAssembly wants to ensure that every function is always invoked with the expected signature.  It has two different instructions for calling a function: a `call` instruction for calling a function whose signature is validated at load time, and a `call_indirect` instruction for calling a function pointer whose signature is validated at runtime. So `call_indirect` is a bit more expensive. If runtime validation fails, then there is a trap.

Section 6.3.2.3, paragraph 8 of the C spec says:

>  A pointer to a function of one type may be converted to a pointer to a function of another type and back again; the result shall compare equal to the original pointer. If a converted pointer is used to call a function whose type is not compatible with the pointed-to type, the behavior is undefined.

But in every architecture + ABI other than wasm, it just works: if you cast a function that takes 2 arguments to a function that takes 3 and then call it with 3 arguments, everything goes fine. In wasm, it traps.

People frequently write getters and setters that don't take a `void* closure` argument and `METH_NOARGS` functions that don't take the always `NULL` second argument. So we need some way to fix it.

### How the JS trampoline works

Point is simply that calling a JS function is relaxed about arguments. If you give a JS function extra arguments, they are just ignored. So we call wasm --> Js --> Wasm. The Wasm --> Js call is strict about wanting 4 arguments (the function pointer and the original three arguments) and the Js --> Wasm call drops any extra arguments.

### What is JSPI (Wasm/JavaScript Promise integration)

The point is that most system calls in Posix are synchronous, whereas most browser APIs are asynchronous. So for instance, people want to use `input()` to get input from the user but the only ways of getting input are async. JSPI allows switching the stack, waiting for user input, then unswitching the stack and continuing. So that stdin can work correctly again. The JSPI is an implementation-stage proposal which allows a very specific sort of stack switching. There is a more general wasm stack switching API which is still under discussion.

### What is the problem

JSPI can only unwind wasm stack frames not JS stack frames. This is because there was concern that the interaction between asyncio unwinding and a second stack switching mechanism could cause subtle bugs in the JS runtime state.

File: Python/emscripten_trampoline.c
Comment (erlend-aasland) [reply Python/emscripten_trampoline.c]: I cannot tell you where, but we can ask @kumaraditya303 or @vstinner :)

Comment (hoodmane) [reply Python/emscripten_trampoline.c]: I stuck it at the end of `pycore_init_runtime()` which feels appropriate to me. LMK if it should go somewhere else.

File: Python/emscripten_trampoline.c
Comment (erlend-aasland) [inline Python/emscripten_trampoline.c]: This should _probably_ go in the interpreter state (`PyInterpreterState`).

**Comment by erlend-aasland** [comment]
@brandtbucher, would you mind taking a look at these two comments?

- https://github.com/python/cpython/pull/106219#discussion_r1247150002
- https://github.com/python/cpython/pull/106219#discussion_r1253620788

File: Python/emscripten_trampoline.c
Comment (hoodmane) [reply Python/emscripten_trampoline.c]: well I think it is not necessary to put it there. I don't see anything in `PyInterpreterState` that is about runtime determination of the architecture. It is okay for this to be a mutable global because it is actually measuring a property of the "hardware" that the Python is running on. So if two interpreters are started and it is initialized in one or both, it will always get assigned the same value both times. As long as the value is filled in at least once, it's okay.

Though maybe it's harmless to put it there. I'm certainly happy either way =)

Comment (erlend-aasland) [reply Python/emscripten_trampoline.c]: Makes sense :)

Comment (kumaraditya303) [reply Python/emscripten_trampoline.c]: You can put in the global runtime struct `_PyRuntime`, we have consolidated all globals there.

File: Python/emscripten_trampoline.c
Comment (kumaraditya303) [reply Python/emscripten_trampoline.c]: LGTM.

File: Python/emscripten_trampoline.c
Comment (hoodmane) [reply Python/emscripten_trampoline.c]: Thanks for the info @kumaraditya303, will do.

**Comment by hoodmane** [comment]
Okay I think I made the changes suggested by @kumaraditya303 and @erlend-aasland. 

Could someone trigger the Emscripten build bot on this PR? If it passes then I think it should be ready to merge (unless of course anyone has further feedback).

**Comment by hoodmane** [comment]
Though this patch is passing the Python core test suite and extra Pyodide tests over here:
https://github.com/pyodide/pyodide/pull/3964
So I think it is okay.

**Comment by erlend-aasland** [comment]
!buildbot .*Emscripten.*

**Comment by erlend-aasland** [comment]
!buildbot .*emscripten.*

**Comment by erlend-aasland** [comment]
BTW; thanks for the explanations, and sorry 'bout the delay (I haven't had bandwidth for CPython the last few weeks.)

**Comment by erlend-aasland** [comment]
@brettcannon, do you want to take a look at this? The changes make sense to me, based on @hoodmane's helpful explanations, but I'm on thin ice here.

File: Python/pystate.c
Comment (erlend-aasland) [inline Python/pystate.c:454]: I'd expect `_initialized = 1` to be the last thing happening in this function, so I'd consider moving `_Py_EmscriptenTrampoline_Init` up a line or two.

File: Python/pystate.c
Comment (erlend-aasland) [inline Python/pystate.c]: Nit: the include list is alphabetically sorted ;)

**Comment by hoodmane** [comment]
Thanks @erlend-aasland. It was actually just as well that it got delayed a bit because it gave me a chance to test it out more carefully and fix a few minor issues. Currently it is still generating some warnings on 64 bit platforms:
```
Warning: ../cpython-ro-srcdir/Objects/descrobject.c:24:26: warning: cast between incompatible function types from ‘getter’ {aka ‘struct _object * (*)(struct _object *, void *)’} to ‘PyObject * (*)(PyObject *, PyObject *, PyObject *)’ {aka ‘struct _object * (*)(struct _object *, struct _object *, struct _object *)’} [-Wcast-function-type]
```
I wonder if CI should fail when there are compile warnings on Ubuntu/Mac/Windows...

File: Python/emscripten_trampoline.c
Comment (erlend-aasland) [inline Python/emscripten_trampoline.c:1]: Nit: I think it would make sense to align the C source sections with PEP-7. I'm not sure if it makes sense to coerce the inlined JS into PEP-7, though :)

File: Python/pylifecycle.c
Comment (erlend-aasland) [inline Python/pylifecycle.c]: Nit: revert spurious change :)

```suggestion
```

File: Python/pystate.c
Comment (hoodmane) [reply Python/pystate.c]: For some reason I was thinking that e came before c. I almost just responded to say that it *is* sorted.

File: Python/emscripten_trampoline.c
Comment (hoodmane) [reply Python/emscripten_trampoline.c:1]: PEP 7 seems to apply perfectly well to the JS code here. I think I correctly applied the PEP 7 guidelines

**Comment by hoodmane** [comment]
Any update on this?

**Comment by erlend-aasland** [comment]
> Any update on this?

I was hoping to get a review from @brettcannon, since he's been involved with the WASM stuff.

**Comment by brettcannon** [comment]
I don't have any opinion on Emscripten stuff (my WebAssembly knowledge is either general WebAssembly or WASI, not the browser stuff). And sorry in the delay; stuff going on in the packaging and SC side of things.

File: Include/internal/pycore_runtime.h
Comment (brettcannon) [inline Include/internal/pycore_runtime.h]: Why does this need to be publicly accessible? And where is it used?

File: Python/pystate.c
Comment (brettcannon) [inline Python/pystate.c:454]: Not seeing this in an `ifdef` guard made me wonder why this check was being done unconditionally? Only after having to follow the code around did I realize it was guarded in an `ifdef` in the header file. I think it would be better, though, to guard it here as well to avoid anyone else having to go digging into the code to wonder what's going on.

**Review by brettcannon — CHANGES_REQUESTED** [review]

File: Include/internal/pycore_runtime.h
Comment (gpshead) [reply Include/internal/pycore_runtime.h]: _(also consider using a bool type from stdbool.h instead of int given this seems to be boolean?)_

**Comment by hoodmane** [comment]
Thanks for the review @brettcannon!

File: Include/internal/pycore_runtime.h
Comment (hoodmane) [reply Include/internal/pycore_runtime.h]: It doesn't need to be publicly accessible. I originally had it as a `static` global in `emscripten_trampoline.c` but I was asked by reviewers to move it here:
https://github.com/python/cpython/pull/106219#discussion_r1253620788
I am perfectly happy to put it back as a `static` global. And yes, it is a boolean I can type it as a `bool`.

Comment (gpshead) [reply Include/internal/pycore_runtime.h]: I much prefer it in runtime state (as you have it) or interpreter state to being a static global.

Comment (hoodmane) [reply Include/internal/pycore_runtime.h]: So it doesn't really need to be public but it is so for tidiness reasons. The variables in this struct aren't super public anyways though? I will add a comment here saying which file uses it.

Comment (Yhg1s) [reply Include/internal/pycore_runtime.h]: Can we put optional bits (and new bits in general) at the end of the struct, instead of near the start? It makes debugging a little easier if the layout doesn't vary as much.

**Comment by hoodmane** [comment]
@brettcannon I think I addressed the recent comments.

File: Include/internal/pycore_runtime.h
Comment (Yhg1s) [inline Include/internal/pycore_runtime.h]:
```suggestion
    int initialized;

```

(Restore the removed empty line.)

**Comment by brettcannon** [comment]
Doing a favour for @hoodmane ...

I have made the requested changes; please review again.

**Comment by brettcannon** [comment]
@hoodmane sorry for not seeing the need for another review! You forgot to leave the magic comment to trigger a re-review or refresh my review, so I didn't realize this still needed anything.

**Comment by brettcannon** [comment]
And FYI there's a merge conflict.

File: Include/internal/pycore_object.h
Comment (brettcannon) [inline Include/internal/pycore_object.h]: I think the CI failures is happening because of this removal.

**Review by brettcannon — CHANGES_REQUESTED** [review]

File: Include/internal/pycore_object.h
Comment (hoodmane) [reply Include/internal/pycore_object.h]: I guess I didn't merge correctly.

**Review by brettcannon — APPROVED** [review]

**Comment by brettcannon** [comment]
!buildbot wasm32-emscripten *

**Comment by brettcannon** [comment]
@hoodmane looks like there were some compiler errors with Emscripten

**Comment by hoodmane** [comment]
Indeed, the failure seems to be:
```
wasm-ld: error: unable to find library -latomic
```
In the configure step I see:
```
checking for stdatomic.h... yes
checking for builtin __atomic_load_n and __atomic_store_n functions... yes
...
checking whether libatomic is needed by <pyatomic.h>... yes
```
so probably we want that last part to say "no" to `libatomic` needed.

**Comment by erlend-aasland** [comment]
> Indeed, the failure seems to be:
> 
> ```
> wasm-ld: error: unable to find library -latomic
> ```

cc. @vstinner (gh-109101, 1f7e42131d2800f0fbb89bfd91fafa8a073e066d)

**Comment by hoodmane** [comment]
Oddly, when I try that locally it seems to work alright.

`test.c`:
```C
// pyatomic.h needs uint64_t and Py_ssize_t types
#include <stdint.h>  // int64_t, intptr_t
#ifdef HAVE_SYS_TYPES_H
#  include <sys/types.h> // ssize_t
#endif
// Code adapted from Include/pyport.h
#if HAVE_SSIZE_T
typedef ssize_t Py_ssize_t;
#elif SIZEOF_VOID_P == SIZEOF_SIZE_T
typedef intptr_t Py_ssize_t;
#else
#  error "unable to define Py_ssize_t"
#endif
#include "cpython/pyatomic.h"
int main()
{
    uint64_t byte;
    _Py_atomic_store_uint64(&byte, 2);
    if (_Py_atomic_or_uint64(&byte, 8) != 2) {
        return 1; // error
    }
    if (_Py_atomic_load_uint64(&byte) != 10) {
        return 1; // error
    }
    return 0; // all good
}
```

```console
$ emcc test.c -IInclude/ && node a.out.js && echo $?
0
```

**Comment by vstinner** [comment]
I don't know why configure considers that this platform needs libatomic. It is the case if the following code fails to build. Can someone try to build it for/on a wasm32 target?

You may have to define/undefine HAVE_SYS_TYPES_H. Or maybe add ``#include "pyconfig.h"`` at the topc which defines or undefines this macro.

```c
// pyatomic.h needs uint64_t and Py_ssize_t types
#include <stdint.h>  // int64_t, intptr_t
#ifdef HAVE_SYS_TYPES_H
#  include <sys/types.h> // ssize_t
#endif
// Code adapted from Include/pyport.h
#if HAVE_SSIZE_T
typedef ssize_t Py_ssize_t;
#elif SIZEOF_VOID_P == SIZEOF_SIZE_T
typedef intptr_t Py_ssize_t;
#else
#  error "unable to define Py_ssize_t"
#endif
#include "cpython/pyatomic.h"
int main()
{
    uint64_t byte;
    _Py_atomic_store_uint64(&byte, 2);
    if (_Py_atomic_or_uint64(&byte, 8) != 2) {
        return 1; // error
    }
    if (_Py_atomic_load_uint64(&byte) != 10) {
        return 1; // error
    }
    return 0; // all good
}
```

**Comment by hoodmane** [comment]
Oh @vstinner the problem is that `ac_cv_libatomic_needed` is unconditionally set to `yes` when cross compiling:
```
  if test "$cross_compiling" = yes
then :
  ac_cv_libatomic_needed=yes
```

**Comment by vstinner** [comment]
Where does it come from, ``ac_cv_libatomic_needed=yes``?

**Comment by erlend-aasland** [comment]
> Where does it come from, `ac_cv_libatomic_needed=yes`?

You set it explicitly in the `AC_RUN_IFELSE` macro. That macro takes up to four params:

1. input -- set to the test program (`AC_LANG_SOURCE`)
2. _action-if-true_ -- set to `ac_cv_libatomic_needed=no`
3. _action-if-false_ -- set to `ac_cv_libatomic_needed=yes`
4. _action-if-cross-compiling_ -- also set to `ac_cv_libatomic_needed=yes`


The argument spec is:

```m4
AC_RUN_IFELSE (input,
               [action-if-true],
               [action-if-false],
               [action-if-cross-compiling = ‘AC_MSG_FAILURE’])
```

**Comment by vstinner** [comment]
So... can we set ``ac_cv_libatomic_needed=no`` for the cross compilation case?

**Comment by erlend-aasland** [comment]
> So... can we set `ac_cv_libatomic_needed=no` for the cross compilation case?

I think that's the best option. It implies that when you're cross compiling, you'll have to explicitly pass `LIBATOMIC` to `configure`. That should probably be documented.

**Comment by hoodmane** [comment]
@erlend-aasland could you retrigger the emscripten build bot?

**Comment by vstinner** [comment]
> I think that's the best option. It implies that when you're cross compiling, you'll have to explicitly pass LIBATOMIC to configure. That should probably be documented.

I wrote PR #109211 to set ``ac_cv_libatomic_needed=no`` when Python is being cross-compiled.

**Comment by hoodmane** [comment]
@vstinner @erlend-aasland @brettcannon could any of you trigger the emscripten buildbot when you get a chance?

**Comment by AlexWaygood** [comment]
!buildbot wasm32-emscripten *

**Comment by vstinner** [comment]
I'm sorry, but I broke wasm32 with my libregrtest refactoring work. I modified the code to spawn test worker processes in a different working directory: the temporary directory created by the main test process.

**Comment by vstinner** [comment]
Wasm32 Python can only be run in the Python source code directory?

**Comment by brettcannon** [comment]
> Wasm32 Python can only be run in the Python source code directory?

Currently the build looks for everything in `lib/` for the stdlib (plus whatever you have to tweak to make `sysconfig` work). You can run it from other places, but you have to map the the code back to `lib/` inside of the WASI runtime.

See https://github.com/python/cpython/blob/fbaf77eb9bd1e6812ebf984d32b29b025cc037d6/Tools/wasm/build_wasi.sh#L28 for a succinct way to run the build (https://github.com/python/cpython/tree/main/Tools/wasm#running documents this).

Long-term I want to freeze the stdlib into the binary so this sort of thing isn't an issue.

**Comment by vstinner** [comment]
> I'm sorry, but I broke wasm32 with my libregrtest refactoring work. I modified the code to spawn test worker processes in a different working directory: the temporary directory created by the main test process.

I wrote PR #109290 to fix Emscripten and WASI buildbot workers.

**Comment by vstinner** [comment]
I merged a first regrtest fix for wasm/wasi: https://github.com/python/cpython/pull/109313#issuecomment-1715704259

Sadly, I didn't notice but my other regrtest json file descriptor change also broke wasm/wasi! Apparently, passing a file descriptor with ``pass_fds=[json_fd]`` doesn't work on these platforms, whereas using a file descriptor as stdout ``stdout=output_fd`` is fine.

**Comment by vstinner** [comment]
> Sadly, I didn't notice but my other regrtest json file descriptor change also broke wasm/wasi! Apparently, passing a file descriptor with pass_fds=[json_fd] doesn't work on these platforms, whereas using a file descriptor as stdout stdout=output_fd is fine.

I wrote PR #109326 to fix it: use a filename, instead of a file descriptor, on WASM/WASI.

**Comment by vstinner** [comment]
Emscripten and WASI buildbot workers are back to green. Sorry for having broken them, I was in the middle of a large refactoring of the regrtest code ;-)

**Comment by brettcannon** [comment]
!buildbot wasm32-emscripten *

**Comment by brettcannon** [comment]
Thanks for the help, everyone!

**Comment by hoodmane** [comment]
Thanks so much @brettcannon @erlend-aasland @vstinner for all your help with this!

**Pull request merged**
